### PR TITLE
Live preview while editing AXAML files

### PR DESCRIPTION
**Pull request by tomlm**
# TLDR: This checkin adds ability to preview consolonia .axaml files while editing them.

**Backstory**
I ended up implementing multiple layered approaches to try and acheive this goal.

# Attempt 1 - try and not have the built-avalonia viewer puke
Basically, the reason the previewer pukes was because of a variety of issues.
* The previewer looks for static AppBuilder function to instantiate your project, and gallery didn't have one.
* we were using Console APIs that don't exist in the context of VS and so it caused the previewer host to blow up.
* our fonts and lines and such are drawn with essentially 1 pixel = 1 char mapping, which doesn't map to an actual pixelated display.
* We don't set default foreground and background, causing nothing to be displayed.

## Changes Made:
* I added Consolonia.Designer assembly
* Added .UseConsoloniaDesigner().  This effectively detects when are app is being loaded by the previewer in design mode **Designer.IsDesignMode** and makes the following tweaks to the environment only when in design mode
1. it loads normal avalonia desktop environment instead of consolonia
2. It registers a DummyConsole which doesn't use Console API
3. It sets styles to set a default Font, backround, foreground 
The effect of this is that Text and colors basically render correctly, but issues are:
* Scaling is way off
* Anything having to do with lines or rectangles are way off (listbox, boxlinebrush, etc.) 
* I tried a bunch of things around scaling, but it was super janky
Screenshot 2024-11-16 191113

So...basically, it's not as shitty as having a broken pane, but not really useful at all.

# Attempt 2 - Avalonia **PreviewControl** displayed via **Design.Preview** 
I had noticed that Avalonia has the following markup:
```xaml
<Design.PreviewWith>
<...alternate markup
</Design.PreviewWith>
```

That got me thinking that I could create Avalonia control which talks to a consolonia app, and if you simply dropped that control in there it would give you the preview.  

## Changes made
* I added a Consolonia.PreviewHost application which allows you to have a live preview of a xaml file. It notices file changes and ssembly changes and reloads itself automatically
* Added Argument to that app (--buffer width height) which basically asks it to dump the PixelBuffer as output
* Added to Consolonia.Designer a custom control **ConsolePreview**.  This guy launches the PreviewHost with --buffer flag, gets the PixelBuffer and renders it as live Avalonia markup as a series of TextBlocks.
* To test this control I created a project **Avalonia.PreviewHost**. This is simply an avalonia app which uses the **ConsolePreview** control and **Consolonia.PreviewHost** to render in a pane.

The Consolonia.PreviewHost and Avalonia.PreviewHost look and work great.
Here's some clips:

## Consolonia.PreviewHost
Consolonia PreviewHost

## Avalonia.PreviewHost
Avalonia PreviewHost

So this is looking super promising. There are no real issues yet except to hook it up. 

Well. 

# 😠😭💔
The avalonia preview window only honors the **Design.WithPreview** tag for **Style** and **Resources**, **_not for user controls_**. It's a couple line change in avalonia to add the support, but it doesn't work out of the box.

## ConsolePreview with Design.WithPreview markup 
To show what it looks like, here's clip where I used **ConsolePreview** control as the WithPreview for a Resource Dictionary.

ConsolePreviewControl
This highlights several pretty serious issues:
1. it's slow as f__k.
2. As soon as you start typing it says "there's an error" 
3. you have to recompile to get it render again. 
4. There still are artifacts. 

This was a pretty despairing point for me, as I was starting to think this was a no-go. I started to think about creating my own VS integration (Which I have done in the past, but it's a painful thing to do.) If only I could just get the console app to work directly in vs.......hey.....wait a  minute.....

# Attempt 3 - Host the console preview app in a terminal window.

With no more code changes, I tried just running the Consolonia.PreviewHost app in a terminal window. The result is actually pretty magic, and it works in both Visual Studio and Visual Studio Code.

## Visual Studio 
With single file preview
Studio-Single

With project preview
Studio-Project

## Visual Studio Code
With single file preview
Code-SingleFile

With project preview
Code-Project

**Pros:**
* It automatically refreshes on build and file change
* You can interact with it
* 100% fidelity, all animations etc work.

**Cons:** 
* ???

## Changes made
none! 🤪

The only issues are:
* it doesn't sync to the current file you are looking at. 
 
At any rate, I think this is pretty respectable solution, just use Consolonia.PreviewHost.exe for live previews, and why not? We *should* be dog fooded our own tool to make tools, no? 

I may at a future point try and create an integration that tightly bounds the active file to the preview window, but for now I'm going to declare victory.

I'm keeping the **Avalonia.PreviewHost** and **PreivewControl** because I spent a lot of time to get them working, and they don't hurt anything. 

The stuff I did to make it so the built-in previewer doesn't barf I think is useful, but again very easy to not use if you don't want it. It's all isolated in the Consolonia.Designer assembly, and so you can ignore it if you don't like it.

BTW: I made Consolonia.PreviewHost a dotnet tool, so you can install it with 

```
dotnet tool install --global --add-source tools\Consolonia.PreviewHost\Bin\Release  Consolonia.PreviewHost
```
OBTW: I also added 2 new color themes: Fluent and Material. They look very similar but I think it looks way better.

## Release Notes

- **New Features**
  - Introduced `DummyConsole` for basic console functionality.
  - Added `ConsolePreview` for rendering console AXAML outputs in Visual Studio.
  - Implemented JSON serialization for `DrawingBoxSymbol`, `Pixel`, `PixelForeground`, and `SimpleSymbol`.
  - New `Fluent` and `Material` themes added for enhanced UI styling.
  
- **Improvements**
  - Enhanced pixel manipulation and rendering logic in `DrawingContextImpl`.
  - Updated project dependencies and improved theme management.
  - Streamlined application startup process with better command-line argument handling.
  
- **Bug Fixes**
  - Refined equality checks and null handling in various classes.
  
- **Documentation**
  - Updated project files and added launch settings for better configuration.

### CODE REVIEW

File: src/Consolonia.Designer/Consolonia.Designer.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        <PackageReference Include="Avalonia" Version="$(AvaloniaVersion)" />
+        <PackageReference Include="Avalonia.FreeDesktop" Version="$(AvaloniaVersion)" />
+        <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Desktop" Version="$(AvaloniaVersion)" />
+        <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Fonts.Inter" Version="$(AvaloniaVersion)" />
+        <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\Consolonia.Core\Consolonia.Core.csproj" />
+      <ProjectReference Condition="'$(Configuration)' == 'Debug'" Include="..\Tools\Consolonia.PreviewHost\Consolonia.PreviewHost.csproj" />
```
[the remark this answers is not in the document]
Comment (tomlm) [reply src/Consolonia.Designer/Consolonia.Designer.csproj:15]: this is not circular, this is by design.  The ConsolePreview control consumes Consolia.PreviewHost. Avalonia.PreviewHost consumes ConsolePreview control in the designer.

File: src/Consolonia.Core/Drawing/DrawingContextImpl.cs [hunk 2]
```diff
@@ -210,7 +210,8 @@ public void DrawGlyphRun(IBrush foreground, IGlyphRunImpl glyphRun)
                 return;
             }
 
-            var shapedBuffer = (ShapedBuffer)glyphRunImpl.GlyphInfos;
+            var shapedBuffer = glyphRunImpl.GlyphInfos as ShapedBuffer;
```
Comment (jinek) [inline src/Consolonia.Core/Drawing/DrawingContextImpl.cs:213]: Cast was better? with "as" there is no difference between null and different type.

File: src/Consolonia.Core/Infrastructure/InputLessDefaultNetConsole.cs [hunk 3]
```diff
@@ -47,21 +35,63 @@ protected InputLessDefaultNetConsole()
...
         {
-            Console.Title = title;
+#pragma warning disable CA1031 // Do not catch general exception types
+            try
+            {
+                Console.Title = title;
+            }
+            catch (Exception)
+            {
+                // ignored
```
Comment (jinek) [inline src/Consolonia.Core/Infrastructure/InputLessDefaultNetConsole.cs:92]: Why do we need to ignore exceptions here?

File: src/Consolonia.Core/Infrastructure/InputLessDefaultNetConsole.cs [hunk 4]
```diff
@@ -47,21 +35,63 @@ protected InputLessDefaultNetConsole()
...
+                        (int left, int top) = Console.GetCursorPosition();
+                        WriteText(
+                            $"{Esc.Foreground(Colors.Transparent)}{Esc.Background(Colors.Transparent)}{TestEmoji}");
+
+                        // TODO, escape sequence
+                        (int left2, _) = Console.GetCursorPosition();
+                        _supportEmoji = left2 - left == 2;
+                        Console.SetCursorPosition(left, top);
+                    }
+                    catch (Exception)
```
Comment (jinek) [inline src/Consolonia.Core/Infrastructure/InputLessDefaultNetConsole.cs:70]: Same question. If for instance OutOfMemoryException happened, why are we assuming emoji are supported?

File: src/Consolonia.Designer/ConsolePreview.cs [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+
+
+            string? xaml;
+#pragma warning disable CA1031 // Do not catch general exception types
+            try
+            {
+                xaml = File.ReadAllText(xamlPath);
+                ArgumentNullException.ThrowIfNull(xaml);
+            }
+            catch (Exception ex)
```
Comment (jinek) [inline src/Consolonia.Designer/ConsolePreview.cs:164]: Should be IOException only probably, or few more, like security exceptions?

File: src/Consolonia.Designer/ConsolePreview.cs [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+                    if (!string.IsNullOrEmpty(line))
+                    {
+                        Debug.WriteLine("BUFFER RECEIVED");
+#pragma warning disable CA1031 // Do not catch general exception types
+                        try
+                        {
+                            var buffer = JsonConvert.DeserializeObject<PixelBuffer>(line)!;
+                            Dispatcher.UIThread.Invoke(() => Content = RenderPixelBuffer(buffer));
+                        }
+                        catch (Exception ex)
```
Comment (jinek) [inline src/Consolonia.Designer/ConsolePreview.cs:242]: DeserializeObject is expected to throw few particular exceptions only

**Comment by goe5mu** [comment]
@tomlm Do we have several tasks in this PR? Should we merge other PRs first?

**Comment by tomlm** [comment]
I think they are pretty independent

File: src/Consolonia.Core/Drawing/DrawingContextImpl.cs (hunk 2 above)
Comment (tomlm) [reply src/Consolonia.Core/Drawing/DrawingContextImpl.cs:213]: the linter was bitching about it.

File: src/Consolonia.Core/Infrastructure/InputLessDefaultNetConsole.cs (hunk 4 above)
Comment (tomlm) [reply src/Consolonia.Core/Infrastructure/InputLessDefaultNetConsole.cs:70]: I removed this, I no longer need to do this.

File: src/Consolonia.Designer/ConsolePreview.cs (hunk 6 above)
Comment (tomlm) [reply src/Consolonia.Designer/ConsolePreview.cs:242]: Changed to JsonException which is base class for a variety of Json Serializer exceptions.

File: src/Consolonia.Designer/ConsolePreview.cs [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+            catch (UnauthorizedAccessException ex)
+            {
+                Content = new TextBlock { Text = $"Unable to access XAML file. {ex.Message}", Foreground = Brushes.Red };
+                return;
+            }
+            catch (IOException ex)
+            {
+                Content = new TextBlock { Text = $"Unable to load XAML file. {ex.Message}", Foreground = Brushes.Red };
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (tomlm) [reply src/Consolonia.Designer/ConsolePreview.cs:173]: actually, it throws IOException not filenotfoundexception

File: src/Consolonia.Core/Drawing/DrawingContextImpl.cs (hunk 2 above)
Comment (jinek) [reply src/Consolonia.Core/Drawing/DrawingContextImpl.cs]: Do we know which rule is that? I will suppress it better.

**Comment by jinek** [comment]
Lets disable CA1303? @tomlm

File: src/Consolonia.Core/Properties/launchSettings.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "Consolonia.Core": {
+      "commandName": "Project",
+      "commandLineArgs": "s:\\github\\consolonia\\src\\consolonia.gallery\\gallery\\galleryviews\\gallerytextlblock.axaml --buffer 80 50"
```
Comment (jinek) [inline src/Consolonia.Core/Properties/launchSettings.json:5]: Do we need this file for non-executable?

File: src/Consolonia.Designer/ConsolePreview.cs [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+                while (true)
+                {
+                    if (_process == null || _process.HasExited)
+                        return;
+
+                    string? line = await _process!.StandardOutput.ReadLineAsync().ConfigureAwait(false);
+                    if (!string.IsNullOrEmpty(line))
+                    {
+                        Debug.WriteLine("BUFFER RECEIVED");
+#pragma warning disable CA1031 // Do not catch general exception types
```
Comment (jinek) [inline src/Consolonia.Designer/ConsolePreview.cs:246]: Disabling CA1031 looks like not required anymore

**Comment by jinek** [comment]
I've merged main branch and fixes few issues. 
But Gallery preview did not work for me with this error:

```
/usr/lib64/dotnet/dotnet exec --runtimeconfig /home/jinek/Repo/Trash/Consolonia/src/Consolonia.Gallery/bin/Debug/net6.0/Consolonia.Gallery.runtimeconfig.json --depsfile /home/jinek/Repo/Trash/Consolonia/src/Consolonia.Gallery/bin/Debug/net6.0/Consolonia.Gallery.deps.json /home/jinek/.nuget/packages/avalonia/11.0.9/buildTransitive/../tools/netcoreapp2.0/designer/Avalonia.Designer.HostApp.dll --transport tcp-bson://127.0.0.1:37515/ --method avalonia-remote /home/jinek/Repo/Trash/Consolonia/src/Consolonia.Gallery/bin/Debug/net6.0/Consolonia.Gallery.dll
Initializing application in design mode
Obtaining AppBuilder instance from Consolonia.Gallery.Program
Sending StartDesignerSessionMessage
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'typeface')
   at System.ArgumentNullException.Throw(String paramName)
   at System.ArgumentNullException.ThrowIfNull(Object argument, String paramName)
   at Consolonia.Themes.TurboVision.Templates.Controls.Helpers.SymbolsControl.set_Text(String value) in /home/jinek/Repo/Trash/Consolonia/src/Consolonia.Themes.TurboVision/Templates/Controls/Helpers/SymbolsControl.cs:line 77
   at CompiledAvaloniaXaml.!AvaloniaResources.XamlClosure_1.Build(IServiceProvider ) in /home/jinek/Repo/Trash/Consolonia/src/Consolonia.Themes.TurboVision/Templates/Controls/Button.axaml:line 27
   at Avalonia.Markup.Xaml.XamlIl.Runtime.XamlIlRuntimeHelpers.<>c__DisplayClass1_0`1.<DeferredTransformationFactoryV2>b__0(IServiceProvider sp)
   at Avalonia.Markup.Xaml.Templates.TemplateContent.Load(Object templateContent)
   at Avalonia.Markup.Xaml.Templates.ControlTemplate.Build(TemplatedControl control)
   at Avalonia.Controls.Primitives.TemplatedControl.ApplyTemplate()
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Controls.Grid.MeasureCell(Int32 cell, Boolean forceInfinityV)
   at Avalonia.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV, Boolean& hasDesiredSizeUChanged)
   at Avalonia.Controls.Grid.MeasureOverride(Size constraint)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Controls.DockPanel.MeasureOverride(Size constraint)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
```

**Comment by jinek** [comment]
> Lets disable CA1303? @tomlm I would do in a separate PR

I've opened a separate PR for this. Will cleanup this code after in that same PR.
For this one the only question from my side is this one: https://github.com/jinek/Consolonia/pull/150#discussion_r1863941963

File: src/Consolonia.Core/Drawing/DrawingContextImpl.cs (hunk 2 above)
Comment (tomlm) [reply src/Consolonia.Core/Drawing/DrawingContextImpl.cs]: I don't.

File: src/Consolonia.Core/Properties/launchSettings.json (hunk 8 above)
Comment (tomlm) [reply src/Consolonia.Core/Properties/launchSettings.json:5]: oh, no, I think we can delete it.  Sorry.

File: src/Consolonia.Core/Drawing/DrawingContextImpl.cs (hunk 2 above)
Comment (jinek) [reply src/Consolonia.Core/Drawing/DrawingContextImpl.cs]: Ok, then next time just ping me please at the moment it happens, would be easier.
I would investigate that.

**Review by jinek — DISMISSED** [review]

**Comment by tomlm** [comment]
GAH. I believed coderabbit that it was dispossable.

**Comment by tomlm** [comment]
I need you to approve again. this is what you already approved,

**Review by jinek — APPROVED** [review]

**Comment by tomlm** [comment]
W00t!
